modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/kg/indexing/query/builder/FilteredQuery.scala

```diff
@@ -0,0 +1,246 @@
+    */
+  def outgoing(thisFilter: Filter, targetFilter: Filter, pagination: Pagination): String = {
+    s"""
+       |SELECT ?total ?s
+       |WITH {
+       |  SELECT ?s
+       |  WHERE {
+       |
+       |?ss ?p ?s .
+       |${buildWhere(thisFilter.expr, "?ss", "this")}
+       |${buildWhere(targetFilter.expr, "?s", "var")}
+       |  }
+       |} AS %resultSet
+       |WHERE {
+       |  {
+       |    SELECT (COUNT(DISTINCT ?s) AS ?total)
+       |    WHERE { INCLUDE %resultSet }
+       |  }
+       |  UNION
+       |  {
+       |    SELECT *
+       |    WHERE { INCLUDE %resultSet }
+       |    LIMIT ${pagination.size}
+       |    OFFSET ${pagination.from}
+       |  }
+       |}""".stripMargin
+  }
+
+  /**
+    * Constructs a SPARQL query based on the provided filters and pagination settings that also computes the total
+    * number of results.
+    *
+    * @param thisFilter   the filter to be applied to select the objects
+    * @param sourceFilter the filter to be applied to filter the subjects
+    * @param pagination   the pagination settings for the generated query
+    */
+  def incoming(thisFilter: Filter, sourceFilter: Filter, pagination: Pagination): String = {
+    s"""
+       |SELECT ?total ?s
+       |WITH {
+       |  SELECT ?s
+       |  WHERE {
+       |
+       |?s ?p ?ss.
+       |${buildWhere(thisFilter.expr, "?ss", "this")}
+       |${buildWhere(sourceFilter.expr, "?s", "var")}
+       |  }
+       |} AS %resultSet
+       |WHERE {
+       |  {
+       |    SELECT (COUNT(DISTINCT ?s) AS ?total)
+       |    WHERE { INCLUDE %resultSet }
+       |  }
+       |  UNION
+       |  {
+       |    SELECT *
+       |    WHERE { INCLUDE %resultSet }
+       |    LIMIT ${pagination.size}
+       |    OFFSET ${pagination.from}
+       |  }
+       |}""".stripMargin
+  }
+
+  private final case class Stmt(stmt: String, filter: String, variable: String)
+
+  private def buildWhere(expr: Expr, subjectVar: String, varPrefix: String): String = {
+    val atomicIdx = new AtomicInteger(0)
+    def nextIdx(): Int =
+      atomicIdx.incrementAndGet()
+
+    // ?s :p ?var
+    // FILTER (?var IN (term1, term2))
+    def inExpr(expr: InExpr): Stmt = {
+      val idx = nextIdx()
+      val variable = s"?${varPrefix}_$idx"
+      val InExpr(path, TermCollection(terms)) = expr
+      val stmt = s"$subjectVar ${path.show} $variable ."
+      val filter = terms.map(_.show).mkString(s"$variable IN (", ", ", ")")
+      Stmt(stmt, filter, variable)
+    }
+
+    // ?s :p ?var .
+    // FILTER ( ?var op term )
+    def compExpr(expr: ComparisonExpr): Stmt = {
+      val idx = nextIdx()
+      val variable = s"?${varPrefix}_$idx"
+      val ComparisonExpr(op, path, term) = expr
+      Stmt(s"$subjectVar ${path.show} $variable .", s"$variable ${op.show} ${term.show}", variable)
+    }
+
+    def fromStmts(op: LogicalOp, stmts: Vector[Stmt]): String = op match {
+      case And =>
+        val select = stmts.map(_.stmt).mkString("\n")
```

Comment:
name of the variable confusing. in SPARQL context, SELECT is a clause.
What abut ""triples"   or "triplesPattern"